File: trunk/net/Ambis1/ambis1.gui/_template/frm/consulta/QueryForm.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace asr.app.pc.template.frm.consulta
{
    public class QueryForm:repsol.forms.template.consulta.QueryForm
    {
        private System.Windows.Forms.Label labError;
    
        public QueryForm():base()
        {
            InitializeComponent();

            this.RememberUserPreferences = true;
            this.tbar.RenderMode = System.Windows.Forms.ToolStripRenderMode.System;

            this.btImprimir.Visible = true;
            this.chkVerFiltro.Visible = false;
            this.btFiltrar.Enabled = false;
            this.ShowIcon = false;
            this.dgConsulta.MouseDoubleClick += new System.Windows.Forms.MouseEventHandler(dgConsulta_MouseDoubleClick);
        }
        public QueryForm(bool soloconsulta):base(soloconsulta)
        {
            InitializeComponent();
            
            this.RememberUserPreferences = true;
            this.tbar.RenderMode = System.Windows.Forms.ToolStripRenderMode.System;

            this.btImprimir.Visible = false;
            this.chkVerFiltro.Visible = false;
            this.btFiltrar.Enabled = false;
            this.ShowIcon = false;
            this.dgConsulta.MouseDoubleClick += new System.Windows.Forms.MouseEventHandler(dgConsulta_MouseDoubleClick);
        }

        void dgConsulta_MouseDoubleClick(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            System.Windows.Forms.DataGridView.HitTestInfo hit = this.dgConsulta.HitTest(e.X, e.Y);
            if ((hit.ColumnIndex != -1) && (hit.RowIndex != -1))
            {
                if (!this.Modal)
                {
                    if (this.OnlyView)
                        this.btConsulta_record();
                    else
                        this.btModificar_record();
                }
                else
                    this.btSeleccionar_record();
            }
        }

        public System.Windows.Forms.ToolStripButton menuFiltrar
        {
            get
            {
                return this.btFiltrar;
            }
        }
        protected override void btFiltrar_record()
        {
            this.chkVerFiltro.Checked = !this.chkVerFiltro.Checked;
        }
        protected override void btImprimir_record()
        {
            try
            {
                this.Cursor = System.Windows.Forms.Cursors.WaitCursor;
                repsol.forms.DynamicReportViewer vVen = new repsol.forms.DynamicReportViewer(this.dgConsulta);
                vVen.Text = this.Text;
                vVen.ShowSelectionColumns = false;
                vVen.RdlCompanyName = "Affinity Staffing";
                vVen.RdlCompanyUrlLogo = "file://" + System.IO.Path.GetFullPath(@"_template/_images/logo.png");
                vVen.Show();
            }
            catch (Exception ex)
            {
                template._common.messages.ShowError(ex);
            }
            finally
            {
                this.Cursor = System.Windows.Forms.Cursors.Default;
            }
        }
        protected override void btImprimirWithSelColumns_record()
        {
            try
            {
                this.Cursor = System.Windows.Forms.Cursors.WaitCursor;
                repsol.forms.DynamicReportViewer vVen = new repsol.forms.DynamicReportViewer(this.dgConsulta);
                vVen.Text = this.Text;
                vVen.ShowSelectionColumns = true;
                vVen.RdlCompanyName = "Affinity Staffing";
                vVen.RdlCompanyUrlLogo = "file://" + System.IO.Path.GetFullPath(@"_template/_images/logo.png");
                vVen.Show();
            }
            catch (Exception ex)
            {
                template._common.messages.ShowError(ex);
            }
            finally
            {
                this.Cursor = System.Windows.Forms.Cursors.Default;
            }
        }
        #region 20080707: Control error consulta

        public void setError(string msgError)
        {
            this.labError.Visible = !string.IsNullOrEmpty(msgError);
            this.labError.Text = msgError;
        }

        #endregion

        private void InitializeComponent()
        {
            System.ComponentModel.ComponentResourceManager resources = new System.ComponentModel.ComponentResourceManager(typeof(QueryForm));
            this.labError = new System.Windows.Forms.Label();
            this.panDetail.SuspendLayout();
            this.panInfo.SuspendLayout();
            ((System.ComponentModel.ISupportInitialize)(this.dvConsulta)).BeginInit();
            this.SuspendLayout();
            // 
            // panDetail
            // 
            this.panDetail.Controls.Add(this.labError);
            this.panDetail.Controls.SetChildIndex(this.labError, 0);
            // 
            // panInfo
            // 
            this.panInfo.Location = new System.Drawing.Point(0, 127);
            // 
            // labNRegistros
            // 
            this.labNRegistros.Text = "0 Registros";
            // 
            // labError
            // 
            this.labError.Dock = System.Windows.Forms.DockStyle.Bottom;
            this.labError.ForeColor = System.Drawing.Color.Red;
            this.labError.Location = new System.Drawing.Point(0, 359);
            this.labError.Name = "labError";
            this.labError.Size = new System.Drawing.Size(792, 62);
            this.labError.TabIndex = 2;
            this.labError.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
            // 
            // QueryForm
            // 
            this.ClientSize = new System.Drawing.Size(792, 573);
            this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            this.Name = "QueryForm";
            this.Load += new System.EventHandler(this.QueryForm_Load);
            this.FormClosing += new System.Windows.Forms.FormClosingEventHandler(this.QueryForm_FormClosing);
            this.Controls.SetChildIndex(this.panHead, 0);
            this.Controls.SetChildIndex(this.panInfo, 0);
            this.Controls.SetChildIndex(this.panDetail, 0);
            this.panDetail.ResumeLayout(false);
            this.panInfo.ResumeLayout(false);
            this.panInfo.PerformLayout();
            ((System.ComponentModel.ISupportInitialize)(this.dvConsulta)).EndInit();
            this.ResumeLayout(false);
            this.PerformLayout();

        }

        private void QueryForm_Load(object sender, EventArgs e)
        {
            try
            {
                if (this.RememberUserPreferences) 
                    loadUserPreferences();

                if (this.Modal)
                {
                    this.btNuevo.Visible = true;
                    this.btModificar.Visible = true;
                    this.btConsulta.Visible = true;

                    this.NuevoToolStripMenuItem.Visible = true;
                    this.ModificarToolStripMenuItem.Visible = true;
                    this.VerToolStripMenuItem.Visible = true;

                    this.DuplicarToolStripMenuItem.Visible = false;
                    this.ToolStripMenuItemSeparador.Visible = false;
                    this.BorrarToolStripMenuItem.Visible = false;
                    this.toolSeparadorBorrar.Visible = false;

                    this.dgConsulta.ContextMenuStrip = this.contextMenu2;
                }

                this.dgConsulta.MouseDown += new System.Windows.Forms.MouseEventHandler(dgConsulta_MouseDown);
            }
            catch (Exception ex)
            {
                template._common.messages.ShowError(ex);
            }
        }

        void dgConsulta_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            try
            {
                System.Windows.Forms.DataGridView dgv = (System.Windows.Forms.DataGridView)sender;
                System.Windows.Forms.DataGridView.HitTestInfo hitTest = dgv.HitTest(e.X, e.Y);
                if ((hitTest.RowIndex != -1) && (hitTest.ColumnIndex != -1))
                    dgv.CurrentCell = dgv[hitTest.ColumnIndex, hitTest.RowIndex];
            }
            catch (Exception ex)
            {
                template._common.messages.ShowError(ex);
            }
        }

        #region "UserPreferences"

        private bool _rememberUserPreferences;

        public new bool RememberUserPreferences
        {
          get { return _rememberUserPreferences; }
          set { _rememberUserPreferences = value; }
        }

        private static System.Collections.Generic.Dictionary<string, repsol.util.setting.userpreferences> _settings;
        private static string _USERPREFERENCES_FULL_PATH = System.IO.Path.GetFullPath(string.Format("{0}{1}{2}", System.Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), System.IO.Path.DirectorySeparatorChar.ToString(), String.Format("{0}{1}.dat", System.Windows.Forms.Application.ProductName, "UserPreferences")));

        public static new string USERPREFERENCES_FULL_PATH
        {
            get { return QueryForm._USERPREFERENCES_FULL_PATH; }
            set { QueryForm._USERPREFERENCES_FULL_PATH = value; }
        }

        private string GetFullPathUserPreferencesDat()
        {
            return USERPREFERENCES_FULL_PATH;
        }

        private void saveUserPreferences()
        {
            System.IO.FileStream fs = null;
            try
            {
                repsol.util.setting.userpreferences currentPreference = getUserPreferences();
                if (currentPreference != null)
                {
                    if (_settings == null) _settings = new Dictionary<string, repsol.util.setting.userpreferences>();
                    if (_settings.ContainsKey(this.GetType().FullName))
                        _settings[this.GetType().FullName] = currentPreference;
                    else
                        _settings.Add(this.GetType().FullName, currentPreference);

                    System.Runtime.Serialization.Formatters.Binary.BinaryFormatter formatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();

                    string fileDatPath = GetFullPathUserPreferencesDat();
                    fs = new System.IO.FileStream(fileDatPath, System.IO.FileMode.Create, System.IO.FileAccess.Write);

                    formatter.Serialize(fs, _settings);
                }
            }
            catch (Exception ex)
            {
                _common.messages.ShowError(ex);
            }
            finally
            {
                if (fs != null) fs.Close();
            }
        }

        private void loadUserPreferences()
        {
            System.IO.FileStream fs = null;
            try
            {
                if (_settings != null)
                {
                    string fileDatPath = GetFullPathUserPreferencesDat();
                    if (System.IO.File.Exists(fileDatPath) && (System.IO.File.ReadAllBytes(fileDatPath).Length > 0))
                    {
                        System.Runtime.Serialization.Formatters.Binary.BinaryFormatter formatter = new System.Runtime.Serialization.Formatters.Binary.BinaryFormatter();

                        fs = new System.IO.FileStream(fileDatPath, System.IO.FileMode.OpenOrCreate, System.IO.FileAccess.Read);

                        _settings = ( Dictionary<string, repsol.util.setting.userpreferences>)formatter.Deserialize(fs);
                    }
                    if ((_settings!=null) && (_settings.ContainsKey(this.GetType().FullName)))
                        setUserPreferences(_settings[this.GetType().FullName]);
                }
            }
            catch (Exception ex)
            {
                _common.messages.ShowError(ex);
            }
            finally
            {
                if (fs != null) fs.Close();
            }
        }

        #endregion

        private void QueryForm_FormClosing(object sender, System.Windows.Forms.FormClosingEventArgs e)
        {
            try
            {
                if ((this.RememberUserPreferences) && (!this.IsDocked))
                    saveUserPreferences();
            }
            catch (Exception ex)
            {
                _common.messages.ShowError(ex);
            }
        }

        
    }
}

File: trunk/net/_template/QueryForm.vb
Namespace frm._template

	Public Class QueryForm
        Inherits repsol.forms.template.consulta.QueryForm

		Private Const STR_Width As String = "Width"
		Private Const STR_Height As String = "Height"
		Private Const STR_Top As String = "Top"
		Private Const STR_Left As String = "Left"
		Protected Friend WithEvents contextMenuAux As System.Windows.Forms.ContextMenuStrip
		Private components As System.ComponentModel.IContainer
		Private _GuardarPosicion As Boolean = True

		Public Sub New()
			MyBase.New()
			InitializeComponent()
			initForm()
		End Sub

		Public Sub New(ByVal soloconsulta As Boolean)
			MyBase.New(soloconsulta)
			InitializeComponent()
			initForm()
		End Sub

		Private Sub InitializeComponent()
            Me.components = New System.ComponentModel.Container
            Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(QueryForm))
            Me.contextMenuAux = New System.Windows.Forms.ContextMenuStrip(Me.components)
            Me.panInfo.SuspendLayout()
            CType(Me.dvConsulta, System.ComponentModel.ISupportInitialize).BeginInit()
            Me.SuspendLayout()
            '
            'panInfo
            '
            Me.panInfo.Location = New System.Drawing.Point(0, 127)
            '
            'contextMenuAux
            '
            Me.contextMenuAux.Name = "contextMenuAux"
            Me.contextMenuAux.Size = New System.Drawing.Size(61, 4)
            '
            'QueryForm
            '
            Me.ClientSize = New System.Drawing.Size(792, 573)
            Me.Icon = CType(resources.GetObject("$this.Icon"), System.Drawing.Icon)
            Me.Name = "QueryForm"
            Me.RememberUserPreferences = True
            Me.Controls.SetChildIndex(Me.panHead, 0)
            Me.Controls.SetChildIndex(Me.panInfo, 0)
            Me.Controls.SetChildIndex(Me.panDetail, 0)
            Me.panInfo.ResumeLayout(False)
            Me.panInfo.PerformLayout()
            CType(Me.dvConsulta, System.ComponentModel.ISupportInitialize).EndInit()
            Me.ResumeLayout(False)
            Me.PerformLayout()

        End Sub

		Protected Overrides Sub btFiltrar_record()
			Me.chkVerFiltro.Checked = Me.btFiltrar.Checked
		End Sub

		Public ReadOnly Property ToolStripMain() As ToolStrip
			Get
				Return tbar
			End Get
		End Property

#Region "Opciones de Botones y Menus contextuales"

        Public Property ContextMenuShow() As Boolean
            Get
                Return Me.contextMenu2.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.contextMenu2.Visible = value
            End Set
        End Property

        Public Property ButtonShowBorrar() As Boolean
            Get
                Return Me.btBorrar.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btBorrar.Visible = value
                Me.BorrarToolStripMenuItem.Visible = Me.btBorrar.Visible
            End Set
        End Property

        Public Property ButtonShowDuplicar() As Boolean
            Get
                Return Me.btDuplicar.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btDuplicar.Visible = value
                Me.DuplicarToolStripMenuItem.Visible = Me.btDuplicar.Visible
            End Set
        End Property

        Public Property ButtonShowFiltro() As Boolean
            Get
                Return Me.btFiltrar.Checked
            End Get
			Set(ByVal value As Boolean)
				Me.chkVerFiltro.Visible = value
				Me.btFiltrar.Checked = value
				Me.chkVerFiltro.Checked = value
			End Set
        End Property

        Public Property ButtonShowModificar() As Boolean
            Get
                Return Me.btModificar.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btModificar.Visible = value
                Me.ModificarToolStripMenuItem.Visible = Me.btModificar.Visible
            End Set
        End Property

        Public Property ButtonShowNuevo() As Boolean
            Get
                Return Me.btNuevo.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btNuevo.Visible = value
                Me.NuevoToolStripMenuItem.Visible = Me.btNuevo.Visible
            End Set
        End Property

        Public Property ButtonShowVer() As Boolean
            Get
                Return Me.btConsulta.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btConsulta.Visible = value
                Me.VerToolStripMenuItem.Visible = value
            End Set
        End Property

        Public Property ButtonShowImprimir() As Boolean
            Get
                Return Me.btImprimir.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btImprimir.Visible = value
                'Me.vVerToolStripMenuItem.Visible = value
            End Set
        End Property

        ' ===========  Obsoletos ==================
        <Obsolete("Evitar su uso. Utilizar ButtonShowVer en su lugar.")> _
        Public Property ShowVerButton() As Boolean
            Get
                Return Me.btConsulta.Visible
            End Get
            Set(ByVal value As Boolean)
                Me.btConsulta.Visible = value
                Me.VerToolStripMenuItem.Visible = value
            End Set
        End Property

        <Obsolete("Evitar su uso. Utilizar ButtonShowFiltro en su lugar.")> _
        Public Property ShowFiltro() As Boolean
            Get
                Return Me.btFiltrar.Checked
            End Get
            Set(ByVal value As Boolean)
                Me.btFiltrar.Checked = value
                Me.chkVerFiltro.Checked = Me.btFiltrar.Checked
            End Set
        End Property

#End Region

#Region "AppSetting"

        Protected Overrides Function getUserPreferences() As repsol.util.setting.userpreferences
            Dim res As New _common.settings.AppPreferences(MyBase.getUserPreferences())
            Return res
        End Function

#End Region

#Region "Gestionar ratón en el grid "

		Private Sub initForm()
			AddHandler Me.dgConsulta.DoubleClick, AddressOf dgConsulta_DoubleClick
			AddHandler Me.dgConsulta.MouseDown, AddressOf dgConsulta_MouseDown
			AddHandler Me.contextMenu2.Opening, AddressOf contextMenu2_opening
		End Sub

		Public Sub dgConsulta_DoubleClick(ByVal sender As Object, ByVal e As EventArgs)
			Dim point As System.Drawing.Point = dgConsulta.PointToClient(System.Windows.Forms.Cursor.Position)
			Dim info As Windows.Forms.DataGridView.HitTestInfo = Me.dgConsulta.HitTest(point.X, point.Y)
			If (info.RowIndex > -1) Then
				If (Me.Modal) Then
					Me.btSeleccionar_record()
				Else
					Me.btConsulta_record()
				End If
			End If
		End Sub	   'Lo sacamos porque en el doble click en el histórico tenemos otra funcionalidad distinta

		Public Sub dgConsulta_MouseDown(ByVal sender As Object, ByVal e As Windows.Forms.MouseEventArgs)
			Dim info As Windows.Forms.DataGridView.HitTestInfo = Me.dgConsulta.HitTest(e.X, e.Y)
			If (info.RowIndex > -1) AndAlso (info.ColumnIndex > -1) Then
				Me.dgConsulta.CurrentCell = Me.dgConsulta.Rows(info.RowIndex).Cells(info.ColumnIndex)
			End If
		End Sub

		Public Sub contextMenu2_opening(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs)
			Dim point As System.Drawing.Point = dgConsulta.PointToClient(System.Windows.Forms.Cursor.Position)
			Dim info As Windows.Forms.DataGridView.HitTestInfo = Me.dgConsulta.HitTest(point.X, point.Y)

			If (Me.dgConsulta.DataSource Is Nothing) OrElse (Me.dgConsulta.Rows.Count = 0) OrElse (info.RowIndex = -1) Then
				e.Cancel = True
			End If
        End Sub

        Protected Sub guardaCAError(ByVal ex As Exception, ByVal frm As repsol.forms.template.BaseForm, Optional ByVal mostrarMensaje As Boolean = True)
            Dim caerror As New VO.caerror.caerrorVO
            caerror.CAEDesError = ex.Message
            caerror.CAEFormulario = frm.Name
            caerror.CAEProcedimiento = ex.TargetSite.Name
            caerror.CAEIdCASesion = _common.objs.CASesion.Sesion.IdCASesion
            Dim accAdd As New BL.caerror.doAdd(caerror)
            accAdd.execute()
            If (mostrarMensaje) Then
                repsol.util.messages.ShowError(ex.Message, frm.Text)
            End If
        End Sub

#End Region

	End Class

End Namespace
